source/common/upstream/conn_pool_map_impl.h

```diff
@@ -7,11 +7,14 @@ namespace Upstream {
 
 template <typename KEY_TYPE, typename POOL_TYPE>
 ConnPoolMap<KEY_TYPE, POOL_TYPE>::ConnPoolMap(Envoy::Event::Dispatcher& dispatcher,
-                                              absl::optional<uint64_t> max_size)
-    : thread_local_dispatcher_(dispatcher), max_size_(max_size) {}
+                                              HostConstSharedPtr host, ResourcePriority priority)
+    : thread_local_dispatcher_(dispatcher), host_(host), priority_(priority) {}
 
-template <typename KEY_TYPE, typename POOL_TYPE>
-ConnPoolMap<KEY_TYPE, POOL_TYPE>::~ConnPoolMap() = default;
+template <typename KEY_TYPE, typename POOL_TYPE> ConnPoolMap<KEY_TYPE, POOL_TYPE>::~ConnPoolMap() {
+  // Explicitly clear things out for resource tracking purposes. Note that we call this rather than
```

Comment:
What's the case where this happens? In general it's better to be consistent on how we delete things vs. have multiple paths. Is it possible to have a single clear path and then have this assert there are no pools?

Reply:
This happens in production when removing a host after draining. https://github.com/envoyproxy/envoy/pull/6298/files#diff-982df999d2cabd23afd6c25088035bf5R901

In testing happens for pretty much every test. It's not the end of the world if we do the deferred delete. Just annoying. :P I'll change it to be consistent. It shouldn't be problematic.

Reply:
I'm struggling with this. Doing the deferred delete makes the integration tests fail catastrophically due to the following assert:

`[2019-03-21 16:32:39.709][16][critical][assert] [source/common/event/dispatcher_impl.cc:45] assert failure: isThreadSafe().`

It's being invoked by the ConnPoolImpl's destructor, whch is in turn being invoked by the dispatcher's destructor.

The stack trace ultimately points us to this:
`Envoy::IntegrationTestServerImpl::createAndRunEnvoyServer()` which I suspect is on a different thread from the actual worker threads, hence the assert failing... I'm not sure offhand what the fix here is. Clear out the deferred delete list prior to stopping?

Reply:
That doesn't work either. :(

If I clean up the deferredDelete list after the TLS has been destroyed, some objects fail destruction because they refer to the TLS -- they trip over an assert. If I clear the deferred delete before that, some of the objects that go onto the deferred delete list as part of destroying the TLS. Basically, they cannot be cleaned up at all.

So, we need to make sure everything cleaned up before the TLS exits -- it needs to happen as part of the TLS exit.

One idea is to explicitly clear the deferred delete list from within the ConnPoolMap:

```  
   clear();

  // Make sure any of the connection pools are immediately cleaned up. There shouldn't be any
  // circular references involved here, so it'll be fine.
  thread_local_dispatcher_.clearDeferredDeleteList();
```

Except this doesn't work. The connection pool itself has a similar strategy: its destructor ends up putting some stuff on the defered delete list, so it tries to clean them up. But, since we're in the middle of a deferred delete, it ignores the current call, deferring it further to the next invocation after the current one completes. So, this works!

```
  clear();

  // Make sure any of the connection pools are immediately cleaned up. There shouldn't be any
  // circular references involved here, so it'll be fine.
  thread_local_dispatcher_.clearDeferredDeleteList();
  thread_local_dispatcher_.clearDeferredDeleteList();
```
But, that's not really an elegant solution. What if the second thing we delete ends up adding another one to the list? This seems pretty fragile.

I don't think it's a good idea to use deferred delete in a destructor unless it's absolutely necessary -- it essentially relies on the dispatcher continuing to run, and if our destructor is going to be invoked during shutdown, we need to be very sure that the deferred delete will "finish" prior to the dispatcher stopping.

So, I'd prefer it if the destructor simply did the book-keeping involved in releasing resources and cleared out the map, leaving the deferredDelete invocation to the call made in the callback.

FWIW, this is the behaviour prior to adding the connection pool map: the hash map mapping keys to connection pools would explicitly destroy the connection pools.

Alternatively, I could make it obvious that there are two clear methods. One which simply clears up the resources (called `clear`), and another which does the deferredDelete as well (`clearWithDifferedDelete`?). The latter would invoke the form. Maybe that would help to make it more explicit that the "clear" path should be common to both.

Reply:
The most recent set of changes take the approach mentioned at the end (with a slightly different name).

Reply:
TBH I'm not crazy about where we ended up here and would prefer what you had previously with a comment, I think it's simpler to understand.

The systemic solution would probably be to modify `clearDeferredDeleteList()` so that it runs in a loop until there is nothing left to delete (basically flipping back and forth between lists), but that's a more invasive/scary change which we should do in a separate PR if we want to go that route. WDYT?

Reply:
I agree -- Loop until it's done would probably fix it. I think it'd handle recursive cases as well (e.g. A does clearDeferredDeleteList which pops on B who does the same, and so on). I also agree that it is better handled in another PR.

I can revert the relevant changes here back to what I had originally.